src/history/hs.c

```diff
@@ -741,14 +736,35 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
         /* Construct the oldest full update. */
         WT_ASSERT(session, modifies.size > 0);
 
-#ifdef HAVE_DIAGNOSTIC
-        __wt_modify_vector_peek(&modifies, &upd);
+        __wt_modify_vector_peek(&modifies, &oldest_upd);
 
-        WT_ASSERT(session, upd->type == WT_UPDATE_STANDARD || upd->type == WT_UPDATE_TOMBSTONE);
-#endif
+        WT_ASSERT(session,
+          oldest_upd->type == WT_UPDATE_STANDARD || oldest_upd->type == WT_UPDATE_TOMBSTONE);
+
+        /*
+         * Clear the history store here if the oldest update is a tombstone and it is the first
+         * update without timestamp on the update chain because we don't have the cursor placed at
+         * the correct place to delete the history store records when inserting the first update and
+         * it may be skipped if there is nothing to insert to the history store.
+         */
+        if (oldest_upd->type == WT_UPDATE_TOMBSTONE && oldest_upd == first_non_ts_upd) {
+            /* We can only delete history store entries that have timestamps. */
+            WT_ERR(__wt_hs_delete_key_from_ts(session, btree->id, key, 1));
+            WT_STAT_CONN_INCR(session, cache_hs_key_truncate_mix_ts);
+            clear_hs = false;
+        } else
+            /*
+             * Clear the content with timestamps in the history store if we see updates without
+             * timestamps on the update chain.
+             *
+             * We don't need to clear the history store records if everything is still on the insert
+             * list and there are no updates moved to the history store by checkpoint or a failed
+             * eviction.
+             */
+            clear_hs = first_non_ts_upd != NULL && !F_ISSET(first_non_ts_upd, WT_UPDATE_HS) &&
+              (list->ins == NULL || ts_updates_in_hs);
```

Comment:
I don't understand `ts_updates_in_hs`. If we've seen a non-timestamped update and it's not the on-page value, doesn't that mean it's going to get written to history store and therefore, will automatically delete entries using the "delete from pos" code path?

Reply:
I changed that. The change is to decide if we need to clear hs in advance and only clear it once in the first insert and avoid inserting anything that is masked by mixed mode operations as they will be deleted anyway.

Reply:
`ts_updates_in_hs` is to see if we have timestamped updates moved to the history store by checkpoint. If we see them, we need to clear hs content.

Reply:
Yeah, the masking part makes sense to me and seems like a good change.

> `ts_updates_in_hs` is to see if we have timestamped updates moved to the history store by checkpoint. If we see them, we need to clear hs content.

Hmm, but if eviction moved them you won't see that right? In that case, we'd wait until we actually inserted the 0 timestamp and then do `__hs_delete_key_from_pos` by doing `next` on the cursor. That method works for both cases so I'm just wondering why we have this separate logic for the case where we can see updates tagged with `WT_UPDATE_HS` in the update list.

Reply:
Checkpoint can move them as well.

Reply:
Yes, I realise that.

My question is why it's necessary in the checkpoint case. After reading more, it seems like it's to avoid work in the case where you're looking at an insert list.

Reply:
Yes, that's correct.